### PR TITLE
Additional filters for a table (from setting)

### LINKED ISSUE #37918: Support post-filters for query tables
> (you don't have to strictly follow this form)

**Use case**

It would be helpful for grafana support. Additional conditions could be applied to specified table.

Example:
```
select sum(x), y from table_a group by y
union all
select a, b from table_b
```
condition: `y > 0 for table_a`

Rewrite a query:
```
select sum(x), y from table_a where y > 0 group by y
union all
select a, b from table_b
```

**Describe the solution you'd like**

A setting with mapping `table_name -> condition_text`.

Implemented in the same way as row-level filter.

**Describe alternatives you've considered**

Exposing an API to modify AST. Would be harder to implement, need maintenance. 

**Additional context**

It's an alternative approach to https://github.com/grafana/clickhouse-datasource/issues/131

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Introduced settings `additional_table_filters`. Using this setting, you can specify additional filtering condition for a table which will be applied directly after reading. Example: `select number, x, y from (select number from system.numbers limit 5) f any left join (select x, y from table_1) s on f.number = s.x settings additional_table_filters={'system.numbers : 'number != 3', 'table_1' : 'x != 2'}`. Introduced setting `additional_result_filter` which specifies additional filtering condition for query result. Closes #37918

### CODE REVIEW

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 1]
```diff
@@ -266,6 +284,71 @@ static void checkAccessRightsForSelect(
...
+    const char * start,
+    const char * end,
+    const DatabaseAndTableWithAlias & target,
+    const Context & context)
+{
+    const char delimiter = ':';
+
+    const char * pos = start;
+    for (; pos < end; ++pos)
+        if (*pos == delimiter)
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterSelectQuery.cpp:299]: This is wrong, the column name can contain `:`

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 2]
```diff
@@ -266,6 +284,71 @@ static void checkAccessRightsForSelect(
...
+    if (setting.empty())
+        return nullptr;
+
+    const char delimiter = ';';
+
+    const char * start = setting.data();
+    const char * end = start + setting.size();
+    for (const char * pos = start; pos < end; ++pos)
+    {
+        if (*pos == delimiter)
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterSelectQuery.cpp:337]: This is wrong. Expression can contain `;`, example:
```
x = ';'
```

File: tests/queries/0_stateless/02346_additional_filters.sql [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+create table m_table (x UInt32, y String) engine = Merge(currentDatabase(), '^table_');
+
+-- { echoOn }
+-- additional filter for inner tables for Merge does not work because it does not create internal interpreter
+-- probably it is expected
+select * from m_table order by x;
+select * from m_table order by x settings additional_filters='table_1 : x != 2';
+select * from m_table order by x  settings additional_filters='table_2 : x != 5';
+select * from m_table order by x  settings additional_filters='table_1 : x != 2; table_2 : x != 5';
+select * from m_table order by x  settings additional_filters='table_1 : x != 4';
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/02346_additional_filters.sql:72]: We also need an additional_filter setting that applies to the resultset.
This is needed for spreadsheet-like UIs.

**Comment by alexey-milovidov** [comment]
I don't like introducing micro-language (serialization/deserialization) inside a setting. Maybe reusing the Map data type will be better?

File: tests/queries/0_stateless/02346_additional_filters.sql (hunk 3 above)
Comment (alexey-milovidov) [reply tests/queries/0_stateless/02346_additional_filters.sql:72]: https://github.com/ClickHouse/ClickHouse/issues/37918#issuecomment-1150544762

File: src/Interpreters/InterpreterSelectQuery.cpp (hunk 1 above)
Comment (KochetovNicolai) [reply src/Interpreters/InterpreterSelectQuery.cpp:299]: it is ok - we always skip only till first `:`.
But if table name can contain `:` then there is a trouble

File: tests/queries/0_stateless/02346_additional_filters.sql (hunk 3 above)
Comment (KochetovNicolai) [reply tests/queries/0_stateless/02346_additional_filters.sql:72]: ok.
maybe there should be a separate setting?
`additional_post_filter`?

still, it is easier to implement on client cause you can use subquery

**Comment by KochetovNicolai** [comment]
I would also prefer Map type. However, it will assume that user will need to escape extra quotes. Like, `additional_filters = '{\'table_1\':\'x != 3\'}'`

**Comment by alexey-milovidov** [comment]
How do we define the Map data type in VALUES? Maybe like `[('table1', 'hello = \'world\'')]`? This is ok.

**Comment by KochetovNicolai** [comment]
Fast tests are ok.
Will not wait for other tests cause last commits did not change any logic

**Comment by UnamedRus** [comment]
Does it work as ROW POLICY or it somehow can be seen in query logs or etc? 

IE it's possible to use it as on fly `row policy` condition.

**Pull request merged**